### ISSUE TITLE
RIOT port B output functionality missing

### ISSUE THREAD

**Issue by ale-79** [JetSetIlly/Gopher2600#17]
Setting RIOT port B as output doesn't seem to be implemented in Gopher2600.

Despite what the "Stella Programmer Guide" says, Port B is fully configurable just like port A, and in fact the 3 unused bit can be used as extra storage if set as output.

More info and a few test roms here:
[https://atariage.com/forums/topic/167173-what-does-the-game-select-switch-actually-do/](https://atariage.com/forums/topic/167173-what-does-the-game-select-switch-actually-do/)

The following roms are the one I posted here
[https://atariage.com/forums/topic/192031-stella-35-released/?do=findComment&comment=2435332](https://atariage.com/forums/topic/192031-stella-35-released/?do=findComment&comment=2435332)
with just a minor change (writes to SWCHA/B now only happen when you actually change the value for those, while before they were written also when you changed SWACNT/SWBCNT)
[riot_port_test_v2.zip](https://github.com/JetSetIlly/Gopher2600/files/7747906/riot_port_test_v2.zip)

In the debugger it would be useful to see both the "Data Register" (in Stella indicated as "SWCHA(W)"/"SWCHB(W)" ) as well as the value that can be read from the port, as they can differ depending on how the "Data Direction Register" (SWACNT/SWBCNT) is set.

**Comment by JetSetIlly** [comment]
Interesting. I'll have a look at that this week. Thanks for the info.

**Comment by JetSetIlly** [comment]
I've pushed a solution to the SWCHB branch.

It passes the tests you posted here and also the tests Omegamatrix posted in the AtariAge thread.

It seems correct but you'll be more attuned to what's correct than me. I'll wait for you approval before merging it into the main branch and closing the issue.

I also still need to add the GUI elements but I'll do that later this week.

**Comment by JetSetIlly** [comment]
I've improved the Ports window (previously called "Chip Registers) so that SWCHB is set according to the value of SWBCNT and SWCHB(W)

The values are updated "live" even if the emulation is not running. For example, if I turn off Bit0 of SWBCNT, the value of SWCHB is updated immediately. Similarly, if I then set Bit0 of SWCHB(W).

(These note also apply to SWCHA of course)

The video below illustrates what I mean.

https://user-images.githubusercontent.com/58750902/147233945-aa209d06-0580-4986-a171-280ec1c2ac6c.mp4

Also, rather than labeling SWCHB with (W) to indicate that this the value set by the CPU, I've chosen to use an icon of a small microchip.

In addition, the "read" SWCHB will show an unlock padlock to show that the value has been altered directly (by the debugger) without consideration of SWCHB(W) or SWBCNT.

I think it works quite well.

The second half of the video shows the GUI updating when the values are set through your test ROM.

Changes pushed to the SWCHB branch
